src/aspire/utils/matrix.py

```diff
@@ -434,6 +434,34 @@ def best_rank1_approximation(A):
     return (U @ S_rank1 @ V).reshape(og_shape)
 
 
+def nearest_rotations(A):
+    """
+    Uses the SVD method to compute the set of nearest rotations to the set A of noisy rotations.
+
+    :param A: A 2D array or a 3D array where the first axis is the stack axis.
+    :return: ndarray of rotations of equal size to A.
+    """
+    og_shape = A.shape
+    dtype = A.dtype
+
+    if A.ndim == 2:
+        A = A[np.newaxis]
+    if A.ndim != 3 or not A.shape[1] == A.shape[2] == 3:
+        raise ValueError(
+            f"Array must be of shape (3, 3) or (n, 3, 3). Found shape {A.shape}."
+        )
+
+    # For the singular value decomposition A = U @ S @ V, we compute the nearest rotation
+    # matrices R = U @ V. If det(U)*det(V) = -1, we negate the third singular value to ensure
+    # we have a rotation.
+    U, _, V = np.linalg.svd(A)
+    neg_det_idx = np.linalg.det(U) * np.linalg.det(V) < 0
+    U[neg_det_idx] = U[neg_det_idx] @ np.diag((1, 1, -1)).astype(dtype, copy=False)
```

Comment:
I think you can just call diag(..., dtype=dtype) here

Reply:
I thought so too, but `diag` doesn't accept a dtype kwarg. It just mutates an array so passes the same dtype. I could do something like `np.diag(np.array([1, 1, -1], dtype=dtype))` if that's preferable.

Reply:
Yeah, you are right.  Nah lets leave it, my mistake!